File: 2642-design-graph-with-shortest-path-calculator/2642-design-graph-with-shortest-path-calculator.py
class Graph(object):

    def __init__(self, n, edges):
        self.node = n
        self.graph = [[] for _ in range(n)]
        for edge in edges:
            self.graph[edge[0]].append((edge[1], edge[2]))
        

    def addEdge(self, edge):
        self.graph[edge[0]].append((edge[1], edge[2]))
        

    def shortestPath(self, start, end):
        distance = [10**9]*self.node
        distance[start] = 0
        pq = [(0, start)] 

        while pq:
            dist, node = heapq.heappop(pq)

            for curr_node, curr_dist in self.graph[node]:
                total_dist = curr_dist + dist
                if distance[curr_node] > total_dist:
                    distance[curr_node] = total_dist
                    heapq.heappush(pq, (total_dist, curr_node))

        print(distance)
        return distance[end] if distance[end] != 10**9 else -1
            

        


# Your Graph object will be instantiated and called as such:
# obj = Graph(n, edges)
# obj.addEdge(edge)
# param_2 = obj.shortestPath(node1,node2)

File: 2642-design-graph-with-shortest-path-calculator/2642-design-graph-with-shortest-path-calculator.java
class Graph {
    private List<List<int[]>> graph;

    public Graph(int n, int[][] edges) {
        graph = new ArrayList<>(n);
        for(int i = 0; i<n; i++) {
            graph.add(new ArrayList<>());
        }

        for(int[] edge : edges) {
            addEdge(edge);
        }
    }
    
    public void addEdge(int[] edge) {
        int [] arr = {edge[1], edge[2]};
        graph.get(edge[0]).add(arr);
    }
    
    public int shortestPath(int start, int end) {
        int n = graph.size();
        int[] distance = new int[n];
        Arrays.fill(distance, Integer.MAX_VALUE);
        distance[start] = 0;

        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(o->o[0]));
        pq.add(new int[]{0, start});

        while(!pq.isEmpty()) {
            int [] current = pq.poll();
            int node = current[1], dist = current[0];

            if(node == end) return dist;

            for(int[] edge : graph.get(node)) {
                int curr_node = edge[0];
                int curr_dist = edge[1] + dist;

                if(curr_dist < distance[curr_node]) {
                    distance[curr_node] = curr_dist;
                    pq.add(new int[] {curr_dist, curr_node});
                }
            }
        }

        return (distance[end] == Integer.MAX_VALUE) ? -1 : distance[end];
    }
}

/**
 * Your Graph object will be instantiated and called as such:
 * Graph obj = new Graph(n, edges);
 * obj.addEdge(edge);
 * int param_2 = obj.shortestPath(node1,node2);
 */